### PR TITLE
Feature/designable migration tool

### PR DESCRIPTION
**Pull request by simonmitchell**
## Description

Adds a CLI tool to handle migrating interface builder files to v2.0.0 of `ThunderBasics` 

## Motivation and Context

Migrating to v2.0.0 will give silent crashes where IBDesignable properties are used in interface-builder. This tool migrates those properties to use pre-existing `CALayer` properties

## How Has This Been Tested?

By running locally on an example project

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: IBMigrationTool/main.swift [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    
+    /// Recurses the given directory for files with a given extension
+    /// - Parameters:
+    ///   - type: The extension to find files for
+    ///   - directory: The directory to search in
+    /// - Returns: An array of file paths
+    func recursivePathsForResource(_ type: String?, directory: String) -> [String] {
+        
+        var filePaths: [String] = []
+        let enumerator = FileManager.default.enumerator(atPath: directory)
```
Comment (BenShutt) [inline IBMigrationTool/main.swift]: As this in an extension of `FileManager` we probably want to remove `FileManager.default.` or make it `static` (but then I might not have it on the `FileManager`)

File: IBMigrationTool/main.swift [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        
+        while let element = enumerator?.nextObject() as? NSString {
```
Comment (BenShutt) [inline IBMigrationTool/main.swift]: 99% sure this is fine but just want to check it's not `URL` (or `NSURL`)

File: IBMigrationTool/main.swift [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    // far simpler so we will leave be for now!
+    string = string.replacingOccurrences(
+        of: "<userDefinedRuntimeAttribute(\\s+)type=\"point\"(\\s+)keyPath=\"shadowOffset\">(\\s+)(<point(\\s)+key=\"value\"(\\s)+x=\"([.\\d]+)\"(\\s+)y=\"([.\\d]+)\"\\/>)",
+        with: "<userDefinedRuntimeAttribute$1type=\"size\"$2keyPath=\"layer.shadowOffset\">$3<size$5key=\"value\"$6width=\"$7\"$8height=\"$9\"\\/>",
+        options: .regularExpression,
+        range: nil
+    )
+    
+    // Save file to disk!
+    let newData = string.data(using: .utf8)
```
Comment (BenShutt) [inline IBMigrationTool/main.swift]: Minor, but I'd move this down and throw some custom `Error` on `nil`

File: IBMigrationTool/main.swift [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        print("Wrote migrated file to \(path)")
+    } catch {
+        print("Failed to write migrated file to \(path)")
+    }
+}
+
+print("This tool will make changes to the Interface Builder (.xib/.storyboard) files in the chosen directory. Please make sure you have no changes in your index before continuing")
+print("Please enter the file path to the Project you want to migrate Interface Builder files to ThunderBasics 3.0.0")
+var filePath = readLine(strippingNewline: true)
+while filePath == nil {
```
Comment (BenShutt) [inline IBMigrationTool/main.swift:61]: Could this go on forever?

File: IBMigrationTool/main.swift [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        )
+    }
+    
+    // shadowOffset is a bit more complex because we used `CGPoint` but
+    // `CALayer` uses `CGSize`
+    
+    // Note this regex isn't perfect as it allows for x,y to take the form 1.2.1.3.1.2 e.t.c
+    // however this should be impossible to enter in Interface Builder and it makes replacing it
+    // far simpler so we will leave be for now!
+    string = string.replacingOccurrences(
```
Comment (BenShutt) [inline IBMigrationTool/main.swift]: This is a long shot and defo not required if it's going to be long, but would be reeeeally useful to see number of changes per file (so we know when nothings changed etc)

Comment (BenShutt) [reply IBMigrationTool/main.swift]: As you point out the git diff can do this for us :)

File: IBMigrationTool/InterfaceBuilderFileMigrator.swift [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        migrateUserDefinedRuntimeAttributes()
+        migrateRemovedCustomClasses()
+    }
+    
+    private func migrateRemovedCustomClasses() {
+        
+        Self.removedCustomClasses.forEach { (customClass) in
+            
+            string = string.replacingOccurrences(
+                of: "\\s+customClass=\"\(customClass)\"\\s?(customModule=\"ThunderBasics\")?",
```
Comment (BenShutt) [inline IBMigrationTool/InterfaceBuilderFileMigrator.swift]: Does the `customModule` always come after the `customClass`? 😄.

Comment (simonmitchell) [reply IBMigrationTool/InterfaceBuilderFileMigrator.swift]: Good question... not entirely sure! Reckon I should add it in the other way too? :)

Comment (BenShutt) [reply IBMigrationTool/InterfaceBuilderFileMigrator.swift]: @simonmitchell I guess there's not really anything to lose by adding it in, performance is not an issue here

**Review by BenShutt — APPROVED** [review]

File: IBMigrationTool/InterfaceBuilderFileMigrator.swift (hunk 6 above)
Comment (simonmitchell) [reply IBMigrationTool/InterfaceBuilderFileMigrator.swift]: Done :D

**Pull request merged**